openhands/runtime/builder/remote.py

```diff
@@ -0,0 +1,111 @@
+import base64
+import io
+import tarfile
+import time
+
+import requests
+
+from openhands.core.logger import openhands_logger as logger
+from openhands.runtime.builder import RuntimeBuilder
+
+
+class RemoteRuntimeBuilder(RuntimeBuilder):
+    """This class interacts with the remote Runtime API for building and managing container images."""
+
+    def __init__(self, api_url: str, api_key: str):
+        self.api_url = api_url
+        self.api_key = api_key
+
+    def build(self, path: str, tags: list[str]) -> str:
+        """Builds a Docker image using the Runtime API's /build endpoint."""
+        # Create a tar archive of the build context
+        tar_buffer = io.BytesIO()
+        with tarfile.open(fileobj=tar_buffer, mode='w:gz') as tar:
+            tar.add(path, arcname='.')
+        tar_buffer.seek(0)
+
+        # Encode the tar file as base64
+        base64_encoded_tar = base64.b64encode(tar_buffer.getvalue()).decode('utf-8')
+
+        # Prepare the multipart form data
+        files = [
+            ('context', ('context.tar.gz', base64_encoded_tar)),
+            ('target_image', (None, tags[0])),
+        ]
+
+        # Add additional tags if present
+        for tag in tags[1:]:
+            files.append(('tags', (None, tag)))
+
+        # Send the POST request to /build
+        headers = {'X-API-Key': self.api_key}
+        response = requests.post(f'{self.api_url}/build', files=files, headers=headers)
+
+        if response.status_code != 202:
+            logger.error(f'Build initiation failed: {response.text}')
+            raise RuntimeError(f'Build initiation failed: {response.text}')
+
+        build_data = response.json()
+        build_id = build_data['build_id']
+        logger.info(f'Build initiated with ID: {build_id}')
+
+        # Poll /build_status until the build is complete
+        while True:
```

Comment:
We should time this out after some reasonable limit

Reply:
done!